File: go/unique.go
package main
import (
  "fmt"
  "sort"
  "strings"
)


func isIn(unique string, c string) bool {
  if len(unique) == 0 {
    return false
  }

  fmt.Println(unique, c)

  for _, s := range unique {
    if string(s) == c {
      return true
    }
  }
  return false
}


// SortString implementation
func SortString(w string) string {
    s := strings.Split(w, "")
    sort.Strings(s)
    return strings.Join(s, "")
}


// O(n^2)
func isUnique1(s string) bool{
  unique := ""

  for i := 0; i < len(s); i++ {
    if !isIn(unique, string(s[i])) {
      unique += string(s[i])
    } else {
      return false
    }
  }
  return true
}


// O(n^2)
func isUnique2(s string) bool{
  for i := 0; i < len(s); i++ {
    for j := i + 1; j < len(s); j++ {
      if s[j] == s[i] {
        return false
      }
    }
  }
  return true
}


// O(n)
func isUnique3(s string) bool{
  uniques := make(map[byte]bool)

  for i := 0; i < len(s); i++ {
      if uniques[s[i]] {
        return false
      }
      uniques[s[i]] = true
  }

  return true
}


// O(n^2*log(n))
func isUnique4(s string) bool{
  sortedString := SortString(s)

  for i := 0; i < len(sortedString) - 1; i++ {
      if sortedString[i] == sortedString[i + 1] {
        return false
      }
    }
  return true
}


func main() {
  str := "vfzabcde"

  fmt.Println(isUnique4(str))
}

File: python/unique.py
# O(n^2)
def is_unique1(string):
    unique = ""

    for c in string: # O(n)
        if c not in unique: # O(n)
            unique += c
        else:
            return False
    return True


# O(n^2)
def is_unique2(string):
    for i, ch1 in enumerate(string): # O(n)
        for ch2 in string[i + 1:]: # O(n)
            if ch1 == ch2:
                return False
    return True


# O(n)
def is_unique3(string):
    uniques_dict = {}

    for c in string: # O(n)
        if uniques_dict.get(c): # O(1)
            return False
        else:
            uniques_dict[c] = True
    return True


# O(n^2*log(n))
def is_unique4(string):
    sorted_string = sorted(string) # O(n*log(n))
    for i in range(len(sorted_string) - 1): #  O(n)
        if sorted_string[i] == sorted_string[i + 1]:
            return False
    return True


def main():
    string = "vfzabcde"
    print(is_unique4(string))


if __name__ == '__main__':
    main()
